### PR TITLE
Annotate collection tree with datasets and cards

### PR DESCRIPTION
**Pull request by dpsutton**
Addresses #18672 

In order to know which collections have items of interest, they are
annotated with two new keys: `:here` and `:below`. These keys will have
a collection (in clojure a set, in js an array) which could contain
dataset and card, depending if they have those items "here" or in
collections below this node.

We include information about which subtrees are interesting so we know
to show or highlight those. We include information about which nodes
have interesting things versus which contain nodes which have things so
we can possibly implement drilldown in an intelligent manner.

ie

```
.
├── root
```

could expand automatically to

```
.
├── root
│   ├── collection
│   │   ├── collection that actually has items
```

without requiring someone to click on the intermediate collection which
has no items itself, just a collection with items.

http://localhost:3000/api/collection/tree?tree=true

```js
{
    "name": "super",
    "id": 154,
    "location": "/",
    "below": ["dataset"]  ;; indicates collections below have datasets
    "children": [{
        "name": "parent",
        "id": 155,
        "location": "/154/",
        "below": ["dataset"]  ;; indicates the same
        "children": [{
            "children": [],
            "slug": "contains",
            "name": "contains",
            "here": ["dataset"],  ;; this node has datasets
            "id": 157,
            "location": "/154/155/",
        }],

    }, {
        "children": [],
        "name": "sibling",
        "id": 156,
        "location": "/154/",
    }],
}
```

### CODE REVIEW

File: test/metabase/models/collection_test.clj [hunk 1]
```diff
@@ -1556,4 +1568,34 @@
...
-               (collection/collections->tree collections)))))))
+               (collection/collections->tree {} collections)))))))
+
+(deftest annotate-collections-test
+  (let [collections [{:id 1, :name "a", :location "/"}
+                     {:id 2, :name "b", :location "/1/"}
+                     {:id 3, :name "c", :location "/1/2/"}
+                     {:id 4, :name "d", :location "/1/2/3/"}
+                     {:id 5, :name "e", :location "/1/"}]
+        clean      #(walk/prewalk
```
Comment (jeff303) [inline test/metabase/models/collection_test.clj:1579]: nit: I think one more space before the `#` ?

File: test/metabase/models/collection_test.clj [hunk 2]
```diff
@@ -1556,4 +1568,34 @@
...
+                     (fn [x]
+                       ;; select important keys and remove empty children
+                       (if (map? x)
+                         (cond-> (select-keys x [:id :here :below :children])
+                           (not (seq (:children x))) (dissoc :children))
+                         x))
+                     %)]
+    (is (= [{:id 1 :name "a" :location "/" :here #{:card} :below #{:card :dataset}}
```
Comment (jeff303) [inline test/metabase/models/collection_test.clj:1587]: nit: other map literals below are using commas to separate kv pairs (sometimes).  This one isn't.

File: test/metabase/models/collection_test.clj [hunk 3]
```diff
@@ -1556,4 +1568,34 @@
...
+            {:id 2, :name "b", :location "/1/" :below #{:dataset}}
+            {:id 3, :name "c", :location "/1/2/" :here #{:dataset} :below #{:dataset}}
+            {:id 4, :name "d", :location "/1/2/3/" :here #{:dataset}}
+            {:id 5, :name "e", :location "/1/" :here #{:card}}]
+           (collection/annotate-collections {:card #{1 5} :dataset #{3 4}} collections)))
+    (is (= [{:id 1 :here #{:card} :below #{:card :dataset}
```
Comment (jeff303) [inline test/metabase/models/collection_test.clj:1593]: nit: slightly hard to follow the structure of this map literal.  Does it make sense to align map values and keep every kv pair on a separate line (i.e. do what's done with `:children` below for every entry)?

**Review by jeff303 — APPROVED** [review]
LGTM with nits

File: test/metabase/models/collection_test.clj (hunk 2 above)
Comment (dpsutton) [reply test/metabase/models/collection_test.clj:1587]: good catch. i'm not really sold on the commas and tend to over look them

File: test/metabase/models/collection_test.clj (hunk 3 above)
Comment (dpsutton) [reply test/metabase/models/collection_test.clj:1593]: Good point. I'll play with the alignment a little. I bias towards code width since i'm split screen buffers all the time. I'll clean it up so its super readable

File: test/metabase/models/collection_test.clj (hunk 1 above)
Comment (dpsutton) [reply test/metabase/models/collection_test.clj:1579]: actually this is properly aligned. The maps visible here are inside of a vector so they are indented one space for the `[` which is in the same column as the `#` for the function literal
